src/prefect/result_records.py

```diff
@@ -0,0 +1,266 @@
+            # and add extra information to help the user recover
+
+            if (
+                isinstance(exc, TypeError)
+                and isinstance(self.result, BaseModel)
+                and str(exc).startswith("cannot pickle")
+            ):
+                try:
+                    from IPython.core.getipython import get_ipython
+
+                    if get_ipython() is not None:
+                        extra_info = inspect.cleandoc(
+                            """
+                            This is a known issue in Pydantic that prevents
+                            locally-defined (non-imported) models from being
+                            serialized by cloudpickle in IPython/Jupyter
+                            environments. Please see
+                            https://github.com/pydantic/pydantic/issues/8232 for
+                            more information. To fix the issue, either: (1) move
+                            your Pydantic class definition to an importable
+                            location, (2) use the JSON serializer for your flow
+                            or task (`result_serializer="json"`), or (3)
+                            disable result persistence for your flow or task
+                            (`persist_result=False`).
+                            """
+                        ).replace("\n", " ")
+                except ImportError:
+                    pass
+            raise SerializationError(
+                f"Failed to serialize object of type {type(self.result).__name__!r} with "
+                f"serializer {self.serializer.type!r}. {extra_info}"
+            ) from exc
+
+        return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def coerce_old_format(cls, value: dict[str, Any] | Any) -> dict[str, Any]:
+        if isinstance(value, dict):
+            if "data" in value:
+                value["result"] = value.pop("data")
+            if "metadata" not in value:
+                value["metadata"] = {}
+            if "expiration" in value:
+                value["metadata"]["expiration"] = value.pop("expiration")
+            if "serializer" in value:
+                value["metadata"]["serializer"] = value.pop("serializer")
+            if "prefect_version" in value:
+                value["metadata"]["prefect_version"] = value.pop("prefect_version")
+        return value
+
+    @classmethod
+    async def _from_metadata(cls, metadata: ResultRecordMetadata) -> "ResultRecord[R]":
+        """
+        Create a result record from metadata.
+
+        Will use the result record metadata to fetch data via a result store.
+
+        Args:
+            metadata: The metadata to create the result record from.
+
+        Returns:
+            ResultRecord: The result record.
+        """
+        # TODO: remove uses of this method; I dont think objects should have classmethods
+        # that elevate them richer versions of themselves
```

Comment:
Can you explain why you think we shouldn't have class methods like this and what you think the alternative should be?

Reply:
Yup, I think methods like this often risk circular imports and excessively tightly coupled code in many cases, e.g.
- it creates unclear object hierarchies (so a future author doesn't perceive any strict rules of what should _depend_ on what)
- objects compiling themselves to actions -> actions that inherit from objects 
- basic metadata objects that have class methods to hydrate rich versions -> the rich versions depends on the metadata objects

And ultimately they're all convenience methods, so I think they should be represented as independent utilities or classmethods on the dependent objects (like `StateCreate.from_state(...)`)

Reply:
Yea, that makes sense - looks like we only use it one place so I'll update in this PR

Reply:
Updated!